File: templates/problems/18_bit_manipulation/03_counting_bits/typescript/countingBits.test.ts
import { describe, it, expect } from 'vitest';
import { countBits } from './countingBits';

describe('CountingBits', () => {
    it('n=2', () => {
        expect(countBits(2)).toEqual([0, 1, 1]);
    });

    it('n=5', () => {
        expect(countBits(5)).toEqual([0, 1, 1, 2, 1, 2]);
    });

    it('n=0', () => {
        expect(countBits(0)).toEqual([0]);
    });

    it('n=1', () => {
        expect(countBits(1)).toEqual([0, 1]);
    });

    it('n=8', () => {
        expect(countBits(8)).toEqual([0, 1, 1, 2, 1, 2, 2, 3, 1]);
    });

    it('n=3', () => {
        expect(countBits(3)).toEqual([0, 1, 1, 2]);
    });
});

File: templates/problems/18_bit_manipulation/03_counting_bits/go/counting_bits_test.go
package counting_bits

import (
	"reflect"
	"testing"
)

func TestCountBits(t *testing.T) {
	tests := []struct {
		name     string
		n        int
		expected []int
	}{
		{"n=2", 2, []int{0, 1, 1}},
		{"n=5", 5, []int{0, 1, 1, 2, 1, 2}},
		{"n=0", 0, []int{0}},
		{"n=1", 1, []int{0, 1}},
		{"n=8", 8, []int{0, 1, 1, 2, 1, 2, 2, 3, 1}},
		{"n=3", 3, []int{0, 1, 1, 2}},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got := countBits(tt.n)
			if !reflect.DeepEqual(got, tt.expected) {
				t.Errorf("countBits(%d) = %v, want %v", tt.n, got, tt.expected)
			}
		})
	}
}
